x-pack/plugins/rule_registry/server/utils/trim_recovered_alerts.ts

```diff
@@ -0,0 +1,47 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import { AlertInstanceContext, AlertInstanceState } from '@kbn/alerting-plugin/common';
+import { PublicAlertFactory } from '@kbn/alerting-plugin/server/alert/create_alert_factory';
+import { ALERT_STATUS, ALERT_STATUS_RECOVERED } from '@kbn/rule-data-utils';
+
+export function trimRecoveredAlerts<
+  InstanceState extends AlertInstanceState = never,
+  InstanceContext extends AlertInstanceContext = never,
+  ActionGroupIds extends string = never
+>(
+  trackedRecoveredEvents: any[],
+  trackedEvents: any[],
+  alertFactory: PublicAlertFactory<InstanceState, InstanceContext, ActionGroupIds>
+) {
+  const recoveredAlerts = [
+    ...trackedRecoveredEvents.map((e, index) => ({
+      index,
+      flappingHistory: e.flappingHistory,
+      trackedEvents: false,
+    })),
+    ...trackedEvents
+      .filter(({ event }) => event[ALERT_STATUS] === ALERT_STATUS_RECOVERED)
+      .map((e, index) => ({ index, flappingHistory: e.flappingHistory, trackedEvents: true })),
+  ];
+
+  const earlyRecoveredAlerts = alertFactory.alertLimit.trimRecovered(recoveredAlerts);
+  // Dropping the "early recovered" alerts for now.
+  // In #143445 we will want to recover these alerts and set flapping to false
+  earlyRecoveredAlerts.forEach((alert) => {
+    if (alert.trackedEvents) {
+      trackedEvents.splice(alert.index as number, 1);
```

Comment:
I thought `trackedEvents` should already be limited by the `alertLimit` in the alert factory since the lifecycle executor reports alerts back to the framework 1-1?

Reply:
I think the limit as applied to the number of tracked events is separate than the number applied to the number of tracked recovered, so if the alert limit is 1000, we can have up to 1000 tracked active alerts and up to 1000 tracked recovered events. That is what I was thinking although I could be wrong!

cc @pmuellr

Reply:
> I think the limit as applied to the number of tracked events is separate than the number applied to the number of tracked recovered, so if the alert limit is 1000, we can have up to 1000 tracked active alerts and up to 1000 tracked recovered events.

Ya, that was my assumption.  The "tracked recovered" alerts were being put in an unbounded container, and we needed a bound, so the thinking was to use the same number.

I think Ying is right that we don't need to trim the tracked events, as that should already be done, but we do need to trim the tracked recovered events here, as that's the other unbounded container used for tracking (for rule registry).

Looking at the implementation for non-rule registry rules, looks like it's doing the same.  Which doesn't seem right.  Since it's potentially removing "recovered" alerts, it's possible then that some alerts would not actually recover.

Reply:
Actually, I wonder if it's even worth the effort to do this work in the rule registry, since this code will end up getting removed when the RR moves to the AF in FAAD.  And there's a separate piece of work that could happen sooner to get rid of the Rule Registry wrapper, which I believe would also make this dead code.

Certainly, if it makes some of this simpler, like not extending the alert methods from the factory, seems like it could be worth it.